src/llama-grammar.cpp

```diff
@@ -682,6 +682,114 @@ static bool llama_grammar_match_partial_char(
     return !is_positive_char;
 }
 
+// transforms a grammar pushdown stack into N possible stacks, all ending
+// at a character range (terminal element)
+// additionally memorizes the stack to its possible stacks by mapping
+// < llama_grammar_stack, llama_grammar_stacks >
+
+struct VectorPointerHash {
+    size_t operator()(const llama_grammar_stack & v) const {
+        size_t seed = v.size();
+        for (const auto* ptr : v) {
+            seed ^= std::hash<const llama_grammar_element*>()(ptr) + 0x9e3779b9 + (seed << 6) + (seed >> 2);
+        }
+        return seed;
+    }
+};
+
+static std::unordered_map<
+    llama_grammar_stack,
+    llama_grammar_stacks,
+    VectorPointerHash>
+    llama_grammar_stacks_cache = {};
+
```

Comment:
It would be necessary to avoid this global state since it is likely not thread-safe

Reply:
Thank you! Would you recommend I update the `advance_stack` header and pass it as a parameter, or should we add it as an attribute in the structure of the `llama_grammar`?